plugins/coverimage.koplugin/main.lua

```diff
@@ -0,0 +1,281 @@
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local lfs = require("libs/libkoreader-lfs")
+local logger = require("logger")
+local util = require("util")
+local _ = require("gettext")
+local T = require("ffi/util").template
+
+local CoverImage = WidgetContainer:new{
+    name = 'coverimage',
+    is_doc_only = true,
+}
+
+function CoverImage:init()
+    self.cover_image_path = G_reader_settings:readSetting("cover_image_path") or "Cover.png"
+    self.cover_image_fallback_path = G_reader_settings:readSetting("cover_image_fallback_path") or "Cover_fallback.png"
+    self.enabled = G_reader_settings:isTrue("cover_image_enabled")
+    self.fallback = G_reader_settings:isTrue("cover_image_fallback")
+    self.ui.menu:registerToMainMenu(self)
+end
+
+function CoverImage:_enabled()
+    return self.enabled
+end
+
+function CoverImage:_fallback()
+    return self.fallback
+end
+
+function CoverImage:dubiousFallbackImage()
+    UIManager:show(InfoMessage:new{
+        text = T(_("\"%1\" \nis not a valid image file!\nPlease correct fallback image in Cover-Image"), self.cover_image_fallback_path),
+        show_icon = true,
+        timeout = 10,
+    })
+end
+
+function CoverImage:showWrongPath( path )
+    UIManager:show(InfoMessage:new{
+        text = T(_("Path of \"%1\" is not accessible.\nPlease correct it."), path),
+        show_icon = true,
+    })
+end
+
+function CoverImage:cleanUpImage()
+    if self.cover_image_fallback_path == "" or not self.fallback then
+        os.remove(self.cover_image_path)
+    elseif lfs.attributes(self.cover_image_fallback_path, "mode") ~= "file" then
+        self:dubiousFallbackImage()
+        os.remove(self.cover_image_path)
+    else
+        os.execute("cp " ..self.cover_image_fallback_path .. " " .. self.cover_image_path)
+    end
+end
+
+function CoverImage:onCloseDocument()
+    logger.dbg("CoverImage: onCloseDocument")
+    if self.fallback then
+        self:cleanUpImage()
+    end
+end
+
+function CoverImage:onReaderReady(doc_settings)
+   logger.dbg("CoverImage: onReaderReady")
+   if self.enabled and not doc_settings:readSetting("exclude_cover_image") == true then
+        local image = self.ui.document:getCoverPageImage()
+        if image then
+            image:writePNG(self.cover_image_path, false)
+        end
+    end
+end
+
+local about_text = _([[
+This plugin saves the current book cover to a file, so it can be used as a screensaver, if your android version and firmware supports it (e.g: Tolinos).
+
+If enabled the cover image of the actual file is stored to the selected screensaver file. Certain books can be excluded.
+
+If fallback is activated, the fallback file will be copied to the screensaver file on book closing.
+If the filename is empty or the file does not exist, the cover file will be deleted and system screensaver is used.
+
+If fallback is not activated the screensaver image remains after closing a book.]])
+
+local function isValidPath(name)
```

Comment:
I do understand what you did here. You're checking if the given path contains a symbolic link, like in /sdcard -> /storage/emulated/0. The right place to do that is in `frontend/device/android/device.lua`, in the specific Device:isValidPath implementation.

That way it can be repurposed for all the rest of the code that uses the function, and here you can do Device:isValidPath(path) directly without this very specific function.

The same with emulator. Right now `isValidPath` returns always true. Your proposal (check if pathExists) makes sense. Do it in `frontend/device/generic/device.lua` in the specific Device:isValidPath implementation.

Reply:
> That way it can be repurposed for all the rest of the code that uses the function, and here you can do Device:isValidPath(path) directly without this very specific function.

Yes, you are right,  I tried to make this plugin as least invasive to the rest of the code as possible. But if you think this change is desirable, I like to put it in `android/device.lua`. But the proposed method is suboptimal. It only works on `/sdcard/` and not on `/sdcard/koreader/`.
I would like to improve the symlink detection on a different PR, 'cause I want this plugin first, and I only want one thing in one PR.

> The same with emulator. Right now `isValidPath` returns always true. Your proposal (check if pathExists) makes sense. Do it in `frontend/device/generic/device.lua` in the specific Device:isValidPath implementation.

Are you sure it is there? I don't find any `isValidPath` with grep on `device/*`, but it's no problem to add it to `generic/device.lua`.

I have done it now, but I like to test the changes before I push it.